kube-proxy can panic if len(HealthzBindAddresses) == 0

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:

`kube-proxy`, from whatever combination of inputs we stuck into it, segfaulted. The segfault, which originates from a slightly modified fork of v1.17, looks as such:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x3d1c786]
goroutine 149 [running]:
k8s.io/kubernetes/pkg/proxy/healthcheck.(*ProxierHealthServer).Updated(0x0)
        /go/src/kubernetes/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/proxy/healthcheck/proxier_health.go:83 +0x26
k8s.io/kubernetes/pkg/proxy/iptables.(*Proxier).SyncLoop(0xc00014db00)
        /go/src/kubernetes/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:474 +0x62
k8s.io/kubernetes/cmd/kube-proxy/app.(*ProxyServer).Run(0xc0001fc480, 0xc0004acc60, 0xc0004acc60)
        /go/src/kubernetes/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kube-proxy/app/server.go:714 +0x735
k8s.io/kubernetes/cmd/kube-proxy/app.(*Options).runLoop.func1(0xc0002b1a80)
        /go/src/kubernetes/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kube-proxy/app/server.go:324 +0x34
created by k8s.io/kubernetes/cmd/kube-proxy/app.(*Options).runLoop
        /go/src/kubernetes/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kube-proxy/app/server.go:323 +0x50
```

Translated to master, where the problem appears to be present, the segfault originates from this call:

https://github.com/kubernetes/kubernetes/blob/d90dd93855ab6c3ccda73dfbd3d19c81f56abca5/pkg/proxy/healthcheck/proxier_health.go#L89

And that is called from here:

https://github.com/kubernetes/kubernetes/blob/d90dd93855ab6c3ccda73dfbd3d19c81f56abca5/pkg/proxy/iptables/proxier.go#L512-L514

This seems odd, because

* The only pointer involved in the segfaulting call is `hs` itself
* The caller nil-checks before calling a method on `hs`.

I did some digging and found #49137, which is a fix for this problem. However, this fix appears to be undone in #83498. There doesn't seem to be a reason for this to have been changed in the latter PR.

**What you expected to happen**:

`kube-proxy` should not segfault from any bizarre combination of user input. If the situation is invalid, it should fail gracefully.

**How to reproduce it (as minimally and precisely as possible)**:

Honestly, I'm not 100% sure what combination of inputs we crammed into `kube-proxy`, but I can locate them and post them if necessary.

**Anything else we need to know?**:

I don't feel confident opening a PR to fix this myself, but in addition to recreating the fix from #49137, a comment should be added explaining why there is this weird, seemingly pointless layer of indirection when assigning to the interface, so that this doesn't get un-fixed again.

--- a/cmd/kube-proxy/app/server.go
+++ b/cmd/kube-proxy/app/server.go
@@ -532,7 +532,7 @@ type ProxyServer struct {
 	UseEndpointSlices      bool
 	OOMScoreAdj            *int32
 	ConfigSyncPeriod       time.Duration
-	HealthzServer          *healthcheck.ProxierHealthServer
+	HealthzServer          healthcheck.ProxierHealthUpdater
 }
 
 // createClients creates a kube client and an event client from the given config and masterOverride.

--- a/cmd/kube-proxy/app/server_others.go
+++ b/cmd/kube-proxy/app/server_others.go
@@ -130,7 +130,7 @@ func newProxyServer(
 		Namespace: "",
 	}
 
-	var healthzServer *healthcheck.ProxierHealthServer
+	var healthzServer healthcheck.ProxierHealthUpdater
 	if len(config.HealthzBindAddress) > 0 {
 		healthzServer = healthcheck.NewProxierHealthServer(config.HealthzBindAddress, 2*config.IPTables.SyncPeriod.Duration, recorder, nodeRef)
 	}

--- a/cmd/kube-proxy/app/server_windows.go
+++ b/cmd/kube-proxy/app/server_windows.go
@@ -92,7 +92,7 @@ func newProxyServer(config *proxyconfigapi.KubeProxyConfiguration, cleanupAndExi
 		Namespace: "",
 	}
 
-	var healthzServer *healthcheck.ProxierHealthServer
+	var healthzServer healthcheck.ProxierHealthUpdater
 	if len(config.HealthzBindAddress) > 0 {
 		healthzServer = healthcheck.NewProxierHealthServer(config.HealthzBindAddress, 2*config.IPTables.SyncPeriod.Duration, recorder, nodeRef)
 	}

--- a/pkg/proxy/healthcheck/proxier_health.go
+++ b/pkg/proxy/healthcheck/proxier_health.go
@@ -42,13 +42,16 @@ type ProxierHealthUpdater interface {
 	// Updated should be called when the proxier has successfully updated the service
 	// rules to reflect the current state.
 	Updated()
+
+	// Run starts the healthz http server and returns.
+	Run()
 }
 
-var _ ProxierHealthUpdater = &ProxierHealthServer{}
+var _ ProxierHealthUpdater = &proxierHealthServer{}
 
-// ProxierHealthServer returns 200 "OK" by default. It verifies that the delay between
+// proxierHealthServer returns 200 "OK" by default. It verifies that the delay between
 // QueuedUpdate() calls and Updated() calls never exceeds healthTimeout.
-type ProxierHealthServer struct {
+type proxierHealthServer struct {
 	listener    listener
 	httpFactory httpServerFactory
 	clock       clock.Clock
@@ -63,12 +66,12 @@ type ProxierHealthServer struct {
 }
 
 // NewProxierHealthServer returns a proxier health http server.
-func NewProxierHealthServer(addr string, healthTimeout time.Duration, recorder record.EventRecorder, nodeRef *v1.ObjectReference) *ProxierHealthServer {
+func NewProxierHealthServer(addr string, healthTimeout time.Duration, recorder record.EventRecorder, nodeRef *v1.ObjectReference) ProxierHealthUpdater {
 	return newProxierHealthServer(stdNetListener{}, stdHTTPServerFactory{}, clock.RealClock{}, addr, healthTimeout, recorder, nodeRef)
 }
 
-func newProxierHealthServer(listener listener, httpServerFactory httpServerFactory, c clock.Clock, addr string, healthTimeout time.Duration, recorder record.EventRecorder, nodeRef *v1.ObjectReference) *ProxierHealthServer {
-	return &ProxierHealthServer{
+func newProxierHealthServer(listener listener, httpServerFactory httpServerFactory, c clock.Clock, addr string, healthTimeout time.Duration, recorder record.EventRecorder, nodeRef *v1.ObjectReference) *proxierHealthServer {
+	return &proxierHealthServer{
 		listener:      listener,
 		httpFactory:   httpServerFactory,
 		clock:         c,
@@ -80,17 +83,17 @@ func newProxierHealthServer(listener listener, httpServerFactory httpServerFacto
 }
 
 // Updated updates the lastUpdated timestamp.
-func (hs *ProxierHealthServer) Updated() {
+func (hs *proxierHealthServer) Updated() {
 	hs.lastUpdated.Store(hs.clock.Now())
 }
 
 // QueuedUpdate updates the lastQueued timestamp.
-func (hs *ProxierHealthServer) QueuedUpdate() {
+func (hs *proxierHealthServer) QueuedUpdate() {
 	hs.lastQueued.Store(hs.clock.Now())
 }
 
 // Run starts the healthz http server and returns.
-func (hs *ProxierHealthServer) Run() {
+func (hs *proxierHealthServer) Run() {
 	serveMux := http.NewServeMux()
 	serveMux.Handle("/healthz", healthzHandler{hs: hs})
 	server := hs.httpFactory.New(hs.addr, serveMux)
@@ -117,7 +120,7 @@ func (hs *ProxierHealthServer) Run() {
 }
 
 type healthzHandler struct {
-	hs *ProxierHealthServer
+	hs *proxierHealthServer
 }
 
 func (h healthzHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {

